### ISSUE TITLE
inplement finite difference WENO schemes

### ISSUE THREAD

**Issue by MilkshakeForReal** [SciML/MethodOfLines.jl#102]
Hi,

I'm wondering if it is possible to add finite volumes for conservative PDEs in the future?

Thank you for your great work btw

**Comment by ChrisRackauckas** [comment]
It would be good to parse the equation and send it to Trixi.jl for semidiscretization when possible.

**Comment by MilkshakeForReal** [comment]
> It would be good to parse the equation and send it to Trixi.jl for semidiscretization when possible.

That is a great package, thanks for the info!

I am trying to use the WENO methods, however. I don't think that is available in `Trixi.jl`.

**Comment by ChrisRackauckas** [comment]
I wonder what @ranocha thinks of a comparison WENO methods.

**Comment by ranocha** [comment]
I like WENO methods in general and it would be great if they were available in Julia. Indeed, we do not have WENO methods (neither finite volume nor finite difference variants) in Trixi.jl since we focus on discontinuous Galerkin methods. That's just the background of our core developers. If someone wanted to implement WENO-type methods, I could give some advice (if desired).
Some aspects to think about
- Finite volume or finite difference style?
- How to handle boundary conditions? There are many different ways to do so, mostly heuristic. Mark Carpenter et al. have some papers with some theoretical foundations, but these methods are of course also more complex (finite difference style).
- Make it general enough to allow many different reconstructions/indicators/fluxes/equations
- Make it general enough to allow well-balanced and positivity preserving methods of Zhang and Shu (shallow water etc.)

**Comment by xtalax** [comment]
Sending to Trixi might have to wait until after we have solution interfaces, as it would affect the way solution retrieval is done in a way that would be confusing, unless there is a way to manually assign observed variables to elements of the `u` array. So that 
```julia
    grid = get_discrete(pdesys, discretization)
    solu = map(d -> sol[d][end], grid[u(x, y, t)])
```
still works, the current recommended practice.

Unless we make dispatch to Trixie a kwarg option which warns and falls back on MOL when a suitable form is not found.

**Comment by MilkshakeForReal** [comment]
A good start would be finite difference [WENO methods](http://www.scholarpedia.org/article/WENO_methods). Then we can move on to reconstruction.

**Comment by MilkshakeForReal** [comment]
A good reference is Section 2.3.4 in 

C.-W. Shu, High order weighted essentially non-oscillatory schemes for convection dominated problems, SIAM Review, 51:82-126, 2009.

**Comment by xtalax** [comment]
Looking at using the scheme presented [here](https://repository.library.brown.edu/studio/item/bdr:297524/PDF/)

See page 8:

Here a scheme is given for `h_i+1/2` , to obtain the scheme for `h_i-1/2` I assume that I reverse the weights and stencils, but what of the `\beta_j` and `\gamma_j` coefficients? 

@ranocha @ChrisRackauckas Do you have an intuition for this?

**Comment by MilkshakeForReal** [comment]
Here is another implementation that might be useful 
https://github.com/tiagopereira/WENO4.jl

**Comment by xtalax** [comment]
Thanks both! I've made a start here.

I have a question about WENO, it is my understanding that there is nothing special about the interpolation weights used, and that others may be substituted as long as they are valid. We have a function in MOL that will generate Lagrange interpolations, automatically adjusting for a nonuniform grid. Would it be valid to use these as long as I calculated the correct form of the beta parameters?
